File: src/main/kotlin/util/Ranker.java
package util;

import main.SpiderMain;
import java.util.*;
import java.util.stream.Collectors;

public class Ranker {
    //Calculate Cosine Similarity of Each Doc

    /*
     *  Weight of query term can be determined by tf * idf
     *
     *  Given:
     *  A query (q) with k query terms  (Let qk denote the weight of term k in the query)
     *  Doc i (di) with k terms         (Let dik denote the frequency of term k in doc i)
     *
     *  Variables needed:
     *  Frequency of query term in the doc  (dik)
     *  Weight of query term                (qk)
     *  Formula:
     *  Sum of all query terms (dik * qk) / sqrt ( (Sum of all query terms in doc (dik))^2 * (Sum of all query terms (qk))^2 )
     */

//    private final String WORD_DB_NAME = "$BASE_URL/rockWord";
//    private final String URL_WORDS_DB_NAME = "$BASE_URL/rockUrlWords";

    //Function to get word ids given an array of strings
    private Vector<String> findWordId(String[] inputWords){
        // word -> wordID
        RocksDB wordDB = new RocksDB(SpiderMain.WORD_DB_NAME);
        Vector<String> result = new Vector<>(0);
        for (String inputWord : inputWords) {
            if (wordDB.get(inputWord) == null){
                //Query term not found in any docs; for now just ignore
                break;
            } else {
                result.add(wordDB.get(inputWord));
            }
        }
        wordDB.close();
//        Collections.sort(result);

        return result;      //Returns a list of wordId that corresponds to query
    }

    //Function to get list of terms from a doc
    private List<String> getTermsFromDoc(String urlKey, RocksDB db){
        // urlID -> list(wordID)
        return CSVParser.INSTANCE.parseFrom(Objects.requireNonNull(db.get(urlKey)));
    }

    //Function to find docs that contains at least one of the query terms
    private Vector<String> findDocs(Vector<String> queryTerms, RocksDB urlWordsDB){
        // urlID -> list(wordID)

        List<String> urlKeys = urlWordsDB.getAllKeys();
        Vector<String> result = new Vector<>(0);

        for (String urlKey : urlKeys) {
            for (String queryTerm : queryTerms) {
                //For each link, check if any of the query term id is present
                if (getTermsFromDoc(urlKey, urlWordsDB).indexOf(queryTerm) != -1) {
                    //Add to doc id to result if query term is in the doc
                    result.add(urlKey);
                    break;
                }
            }
        }

        return result;

    }

    //Given termId and urlId count number of terms in doc
    private int countTermInDoc (String wordIdIn, String urlIdIn, RocksDB urlWordsDB){
        List<String> wordIdsInDoc = getTermsFromDoc(urlIdIn, urlWordsDB);
        int count = 0;

        for (String wordId : wordIdsInDoc){
            if (wordId.equals(wordIdIn))
                count++;
        }

        return count;
    }

    //Returns query weight of term given a term id
    //For now, just return 1.0 every time.
    private double getQueryWeight(String termId){
        return 1.0;
    }

    //Given urlId and queryTermId calculate cosine similarity
    private double cosSim(String urlId, Vector<String> queryTermIds, RocksDB urlWordsDB){
        //dikQk stores Sum of all query terms (dik * qk) [Nominator of cosSim]
        double dikQk = 0.0;

        //dik stores sum of all term frequency in doc squared
        double dik = 0.0;

        //qk stores sum of all query term squared
        double qk = 0.0;

        for (String queryTermId : queryTermIds){
            dikQk += countTermInDoc(queryTermId, urlId, urlWordsDB) * getQueryWeight(queryTermId);
            dik += Math.pow(countTermInDoc(queryTermId, urlId, urlWordsDB), 2);
            qk += Math.pow(getQueryWeight(queryTermId), 2);
        }
        return dikQk / Math.sqrt(dik * qk);

    }

    /*
     * Given an array of queryTerms, return a list of DocIds containing
     * at least one of the query terms, sorted by descending order of cosSim values
     */
//    public List<String> rankDocs(String[] queryTerms, RocksDB urlWordsDB) {

    /*
     * Given an array of queryTerms, return a map of DocIds to score.
     * Contains docs with at least one of the query terms,
     * sorted by descending order of cosSim values
     */
    public Map<String, Double> rankDocs(String[] queryTerms, RocksDB urlWordsDB) {
        Vector<String> queryTermIds = findWordId(queryTerms);
        System.out.println(queryTermIds);
        Vector<String> docsWithQuery = findDocs(queryTermIds, urlWordsDB);
        System.out.println(docsWithQuery);

        Map<String, Double> docCosSim = new HashMap<>();

        for (String docId : docsWithQuery) {
            docCosSim.put(docId, cosSim(docId, queryTermIds, urlWordsDB));
        }

        //Sort by values (cosine similarities)
        Map<String, Double> sortedCosSim = docCosSim
                .entrySet()
                .stream()
                .sorted(Collections.reverseOrder(Map.Entry.comparingByValue()))
                .collect(
                        Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue,
                                (e1, e2) -> e2, LinkedHashMap::new));

        return sortedCosSim;

//        return new ArrayList<>(sortedCosSim.keySet());    //Return only urlId ranked by score
    }


}

File: src/main/kotlin/util/Ranker.kt
package util

import org.apache.commons.lang3.StringUtils
import main.SpiderMain
import java.lang.StringBuilder
import java.util.*

/**
 *  Weight of query term can be determined by tf * idf
 *
 *  Given:
 *  A query (q) with k query terms  (Let qk denote the weight of term k in the query)
 *  Doc i (di) with k terms         (Let dik denote the frequency of term k in doc i)
 *
 *  Variables needed:
 *  Frequency of query term in the doc  (dik)
 *  Weight of query term                (qk)
 *  Formula:
 *  Sum of all query terms (dik * qk) / sqrt ( (Sum of all query terms in doc (dik))^2 * (Sum of all query terms (qk))^2 )
 */
object Ranker {
    private val tfIdfDB = RocksDB(SpiderMain.TF_IDF_DB_NAME)
    private val urlLengthDB = RocksDB(SpiderMain.URL_LENGTH_DB_NAME)
    private val urlWordsDB = RocksDB(SpiderMain.URL_WORDS_DB_NAME)
    private val spiderDB = RocksDB(SpiderMain.SPIDER_DB_NAME)
    private val wordDB = RocksDB(SpiderMain.WORD_DB_NAME)
    private val urls = urlLengthDB.getAllKeys()
    private val tfIdfMap = mutableMapOf<String, SortedMap<String, Double>>()

    init {
        val keys = tfIdfDB.getAllKeys()
        keys.forEach {
            val tfIdfList = CSVParser.parseFrom(tfIdfDB[it] ?: "")
            val sortedMap = sortedMapOf<String, Double>()
            for (i in 0 until tfIdfList.size step 2) {
                sortedMap[tfIdfList[i]] = tfIdfList[i + 1].toDouble()
            }
            tfIdfMap[it] = sortedMap
        }
    }

    fun initialize() {
        println("Total ${urls.size} URLs found in Database")
    }

    fun rankDocs(queryTerms: List<List<String>>): MutableMap<String, Double> {
        val queryTermIds = findWordId(queryTerms, wordDB)
        val resultMap = mutableMapOf<String, Double>()
        queryTermIds.forEach { queryTermId ->
            if (queryTermId.size > 1) {  //Phrase detected
                urls.forEach { urlId ->
                    //Phrase is in document
                    val sb = StringBuilder()
                    for (i in 0 until queryTermId.size) {
                        sb.append("\"${queryTermId[i]}\"")
                        if (i != queryTermId.size - 1) {
                            sb.append(",")
                        }
                    }

                    val queryStr = sb.toString()
                    val wordsInDoc = urlWordsDB[urlId]!!
                    if (wordsInDoc.contains(queryStr)) {
                        //tfIdf: wordId, score, wordId, score...
                        var score = 0.0

                        //Get tfidf score for each word in phrase
                        for (term in queryTermId) {
                            score += tfIdfMap[urlId]!![term]!!
                        }

                        //Multiply score by number of times the phrase appear in doc
                        resultMap[urlId] = (resultMap[urlId] ?: 0.0) +
                                score * StringUtils.countMatches(wordsInDoc, queryStr)
                    }
                }
            } else {            //Single word
                val startTime = System.currentTimeMillis()
                val spiderList = CSVParser.parseFrom(spiderDB[queryTermId[0]]!!)
                println("Term ${queryTermId[0]} parsing took with ${spiderList.size} urls found ${(System.currentTimeMillis() - startTime) / 1000.0} seconds")
                spiderList.parallelStream().forEach { docId ->
                    resultMap[docId] = (resultMap[docId] ?: 0.0) + tfIdfMap[docId]!![queryTermId[0]]!!
                }
                println("Term ${queryTermId[0]} took ${(System.currentTimeMillis() - startTime) / 1000.0} seconds")
            }
        }

        val queryLength = Math.sqrt(queryTermIds.flatten().size.toDouble())
        resultMap.forEach { key, value ->
            resultMap[key] = value / (urlLengthDB[key]!!.toDouble() * queryLength)
        }

        return resultMap
    }

    private fun findWordId(inputWordsList: List<List<String>>, wordDB: RocksDB): List<List<String>> {
        val result = mutableListOf<List<String>>()

        for (inputWords in inputWordsList) {
            val phraseList = mutableListOf<String>()
            for (inputWord in inputWords) {
                val wordId = wordDB[inputWord]
                if (wordId != null) phraseList.add(wordId)
            }
            result.add(phraseList)
        }

        return result
    }
}